File: 000-assignments/02-advanced-dsa/064-linked-list-1-introduction/linkedlist.py
class LinkedList:
    val = None
    next = None

    def __init__(self, val):
        self.val = val

    def append(self, val):
        if self.next is None:
            self.next = LinkedList(val)
        else:
            self.next.append(val)

    def __str__(self):
        return str(self.val) + " -> " + str(self.next)

    def search(self, key):
        if self == None:
            return False

        temp = self
        while temp.next != None:
            if temp.val == key:
                return True
            temp = temp.next
        return False

    def insert(self, val, pos):
        newNode = LinkedList(val)
        if pos == 0:
            newNode.next = self
            return newNode

        temp = self
        index = 1
        while temp.next != None:
            if pos == index:
                newNode.next = temp.next
                temp.next = newNode
                return self
            temp = temp.next
            index += 1

        temp.next = newNode
        return self

    def delete(self, val):
        if self.val == val:
            self = self.next
            return self

        temp = self
        while temp.next != None:
            if temp.next.val == val:
                temp.next = temp.next.next
                return self
            temp = temp.next
        return self

    def reverse(self):
        prev = None
        curr = self
        while curr != None:
            forw = curr.next
            curr.next = prev
            prev = curr
            curr = forw

        self = prev
        return self

    def is_palindrome(self):
        p1 = self
        p2 = self

        n = 0
        temp = self
        while temp != None:
            n += 1
            temp = temp.next

        mid = n // 2
        temp = self
        while mid > 0:
            p2 = p2.next
            mid -= 1
        print(p1, p2)

        p2 = p2.reverse()
        print(p1, p2)

        while p2 != None:
            if p1.val != p2.val:
                return False
            p1 = p1.next
            p2 = p2.next

        return True


if __name__ == "__main__":
    ll = LinkedList(1)
    ll.append(2)
    ll.append(3)
    ll.append(4)
    ll.append(5)
    print(ll)
    # 1 -> 2 -> 3 -> 4 -> 5 -> None

    print("3 is present? ", ll.search(3))
    # True
    print("6 is present? ", ll.search(6))
    # False

    ll = ll.insert(0, 0)
    print(ll)

    ll = ll.insert(6, 6)
    print(ll)

    print(ll)
    ll = ll.insert(3, 3)
    print(ll)

    ll = ll.delete(0)
    print(ll)

    ll = ll.delete(6)
    print(ll)

    ll = ll.delete(3)
    print(ll)

    ll = ll.reverse()
    print(ll)

    ll = ll.reverse()
    print(ll)

    print(ll.is_palindrome())

    ll = LinkedList(1)
    ll.append(2)
    ll.append(3)
    ll.append(2)
    ll.append(1)
    print(ll)

    print(ll.is_palindrome())

    ll = LinkedList(1)
    ll.append(2)
    ll.append(3)
    ll.append(3)
    ll.append(2)
    ll.append(1)
    print(ll)

    print(ll.is_palindrome())

File: 000-assignments/02-advanced-dsa/064-linked-list-1-introduction/linkedlist.go
package main

import (
	"fmt"
	"linkedlist"
)

func main() {
	ll := linkedlist.NewNode(1, 2, 3, 4, 5, 6)
	fmt.Println(ll.String())

	fmt.Println(search(ll, 3))
	fmt.Println(search(ll, 7))

	ll = insert(ll, 0, 10)
	fmt.Println(ll.String())

	ll = insert(ll, 3, 20)
	fmt.Println(ll.String())

	ll = delete(ll, 10)
	fmt.Println(ll.String())

	ll = delete(ll, 20)
	fmt.Println(ll.String())

	ll = reverse(ll)
	fmt.Println(ll.String())
}

func search(ll *linkedlist.Node, k int) bool {

	if ll == nil {
		return false
	}

	temp := ll

	for temp.Next != nil {

		if temp.Value == k {
			return true
		}
		temp = temp.Next
	}
	return false
}

func insert(ll *linkedlist.Node, k int, v int) *linkedlist.Node {

	newLL := linkedlist.NewNode(v)

	if k == 0 {
		newLL.Next = ll
		ll = newLL
	} else {

		temp := ll
		for i := 0; i < k-1; i++ {
			temp = temp.Next
		}

		newLL.Next = temp.Next
		temp.Next = newLL
	}
	return ll
}

func delete(ll *linkedlist.Node, k int) *linkedlist.Node {

	if ll == nil {
		return ll
	}

	temp := ll
	if temp.Value == k {
		ll = temp.Next
		return ll
	}

	for temp != nil {
		if temp.Next != nil && temp.Next.Value == k {
			temp.Next = temp.Next.Next
			return ll
		}
		temp = temp.Next
	}
	return ll
}

func reverse(ll *linkedlist.Node) *linkedlist.Node {

	var prev, curr *linkedlist.Node = nil, ll

	fmt.Println(min(1, 2))

	for curr != nil {

		forward := curr.Next
		curr.Next = prev
		prev = curr
		curr = forward
	}
	ll = prev
	return ll
}
